politeiad/backend/gitbe/cms.go

```diff
@@ -0,0 +1,1183 @@
+	}
+
+	return nil
+}
+func (g *gitBackEnd) cmsPluginJournalFlusher() {
+	// XXX make this a single PR instead of 2 to save some git time
+	err := g.flushDCCVoteJournals()
+	if err != nil {
+		log.Errorf("cmsPluginVoteFlusher: %v", err)
+	}
+}
+
+func (g *gitBackEnd) pluginStartDCCVote(payload string) (string, error) {
+	vote, err := cmsplugin.DecodeStartVote([]byte(payload))
+	if err != nil {
+		return "", fmt.Errorf("DecodeStartVote %v", err)
+	}
+
+	// Verify vote bits are somewhat sane
+	for _, v := range vote.Vote.Options {
+		err = _validateCMSVoteBit(vote.Vote.Options, vote.Vote.Mask, v.Bits)
+		if err != nil {
+			return "", fmt.Errorf("invalid vote bits: %v", err)
+		}
+	}
+
+	// Verify dcc exists
+	tokenB, err := util.ConvertStringToken(vote.Vote.Token)
+	if err != nil {
+		return "", fmt.Errorf("ConvertStringToken %v", err)
+	}
+	token := vote.Vote.Token
+
+	if !g.vettedPropExists(token) {
+		return "", fmt.Errorf("unknown proposal: %v", token)
+	}
+
+	// Make sure vote duration is within min/max range
+	// XXX calculate this value for testnet instead of using hard coded values.
+	if vote.Vote.Duration < cmsplugin.VoteDurationMin ||
+		vote.Vote.Duration > cmsplugin.VoteDurationMax {
+		// XXX return a user error instead of an internal error
+		return "", fmt.Errorf("invalid duration: %v (%v - %v)",
+			vote.Vote.Duration, cmsplugin.VoteDurationMin,
+			cmsplugin.VoteDurationMax)
+	}
+
+	// 1. Get best block
+	bb, err := bestBlock()
+	if err != nil {
+		return "", fmt.Errorf("bestBlock %v", err)
+	}
+	if bb.Height < uint32(g.activeNetParams.TicketMaturity) {
+		return "", fmt.Errorf("invalid height")
+	}
+	// 2. Subtract TicketMaturity from block height to get into
+	// unforkable teritory
+	startVoteBlock, err := block(bb.Height)
+	if err != nil {
+		return "", fmt.Errorf("bestBlock %v", err)
+	}
+
+	// Convert the user weights array into a []string array.
+	// This will become the format "userid+weight"
+	eligbleUsers := make([]string, len(vote.UserWeights))
+	for i, uw := range vote.UserWeights {
+		fmt.Println(i, uw, len(vote.UserWeights))
+		eligbleUsers[i] = uw.UserID + "," + strconv.Itoa(int(uw.Weight))
+	}
+
+	svr := cmsplugin.StartVoteReply{
+		Version:          cmsplugin.VersionStartVoteReply,
+		StartBlockHeight: startVoteBlock.Height,
+		StartBlockHash:   startVoteBlock.Hash,
+		EndHeight:        startVoteBlock.Height + vote.Vote.Duration,
+		EligibleUsers:    eligbleUsers,
+	}
+	svrb, err := cmsplugin.EncodeStartVoteReply(svr)
+	if err != nil {
+		return "", fmt.Errorf("EncodeStartVoteReply: %v", err)
+	}
+
+	// Add version to on disk structure
+	vote.Version = cmsplugin.VersionStartVote
+	voteb, err := cmsplugin.EncodeStartVote(*vote)
+	if err != nil {
+		return "", fmt.Errorf("EncodeStartVote: %v", err)
+	}
+
+	// Verify proposal state
+	g.Lock()
+	defer g.Unlock()
+	if g.shutdown {
+		// Make sure we are not shutting down
+		return "", backend.ErrShutdown
+	}
+	_, err1 := os.Stat(pijoin(joinLatest(g.vetted, token),
```

Comment:
Plugins shouldn't be reading directly from the filesystem for gitbe records.  #1160 added api to prevent this and updated decred plugin to remove the direct filesystem touches, except for journals which is fine for now.  Use the following methods instead:

`g.UnvettedExists()`
`g.VettedExists()`
`g.vettedMetadataStreamExists()`
`g.getVettedMetadataStream()`